File: Java/LinkedListPractice/src/LinkedList/LinkedList.java
package LinkedList;

import java.util.Optional;

public class LinkedList {
    Node head;
    Node tail;

    int maxSize;

    public LinkedList(int maxSize) {
        this.maxSize = maxSize;
    }

    public void add(int data) {
        Node node = new Node(data);
        if (this.head == null) {
            this.head = node;
        } else {
            this.tail.next = node;
        }
        this.tail = node;
    }

    public boolean contains(int data) {
        Node currentNode = this.head;
        while(currentNode != null) {
            if (currentNode.getData() == data) {
                return true;
            }
            currentNode = currentNode.getNext();
        }
        return false;
    }

    public Optional<Integer> get(int index) {
        Node currentNode = this.head;
        if (currentNode != null && index >= 0) {
            int currentIndex = 0;
            while(currentNode != null && currentIndex < index) {
                if(currentNode.getNext() != null) {
                    currentNode = currentNode.getNext();
                    currentIndex++;
                } else {
                    return Optional.empty();
                }
            }
            assert currentNode != null;
            return Optional.of(currentNode.getData());
        }
        return Optional.empty();
    }

    public void insertAt(int index, int data) {
        Node newNode = new Node(data);
        if (index == 0) {
            newNode.setNext(this.head);
            this.head = newNode;
        } else {
            Node previousNode = null;
            Node currentNode = this.head;
            int currentIndex = 0;
            while(currentIndex < index) {
                previousNode = currentNode;
                currentNode = currentNode.getNext();
                currentIndex++;
            }
            newNode.setNext(currentNode);
            assert previousNode != null;
            previousNode.setNext(newNode);
        }
    }

    public void printAll() {
        Node currentNode = this.head;
        while(currentNode != null) {
            System.out.println(currentNode.getData());
            currentNode = currentNode.getNext();
        }
    }

    public void removeFrom(int index) {
        if (index == 0) {
            if (this.head.getNext() != null) {
                this.head = this.head.getNext();
            } else {
                this.head = null;
            }

        } else {
            Node previousNode = null;
            Node currentNode = this.head;
            int currentIndex = 0;
            while(currentIndex < index) {
                previousNode = currentNode;
                currentNode = currentNode.getNext();
                currentIndex++;
            }
            assert previousNode != null;
            previousNode.setNext(currentNode.getNext());
        }
    }
}

File: Kotlin/LinkedListPractice/src/linkedList/LinkedList.kt
package linkedList

class LinkedList(
    var head : Node?,
    var tail : Node?,
    //val maxLength : Int
) {

    fun add(data : Int) {
        val newNode = Node(data=data, next=null)
        if (head == null) {
            head = newNode
        } else {
            tail?.next = newNode
        }
        tail = newNode
    }

    fun contains(data: Int): Boolean {
        var currentNode = head
        while (currentNode != null) {
            if (currentNode.data == data) {
                return true
            }
            currentNode = currentNode.next
        }
        return false
    }

    fun get(data: Int): Node? {
        var currentNode = head
        while (currentNode != null) {
            if (currentNode.data == data) {
                return currentNode
            }
            currentNode = currentNode.next
        }
        return null
    }

    fun insertAt(index : Int, data : Int) {
        println("Insert at method called with index ($index) and data ($data)")
        val newNode = Node(data=data, next=null)
        if (index == 0) {
            newNode.next = head
            head = newNode
        } else {
            var currentIndex = 0
            var currentNode = head
            var previousNode : Node? = null
            while (currentIndex < index && currentNode != null) {
                previousNode = currentNode
                currentNode = currentNode.next
                currentIndex++
            }
            if (currentIndex == index) {
                newNode.next = currentNode
                previousNode!!.next = newNode
            }
        }
    }

    fun length(): Int {
        var length = 0
        var currentNode = head
        while(currentNode != null) {
            length++
            currentNode = currentNode.next
        }
        return length
    }

    fun printAll() {
        var currentNode = head
        while(currentNode != null) {
            println(currentNode.data)
            currentNode = currentNode.next
        }
    }

    fun remove(index : Int) {
        println("Remove method called with index ($index)")
        if (index == 0) {
            head = if (head?.next != null) {
                head!!.next
            } else {
                null
            }
        } else {
            var currentIndex = 0
            var currentNode = head
            var previousNode : Node? = null
            while (currentIndex < index && currentNode != null) {
                previousNode = currentNode
                currentNode = currentNode.next
                currentIndex++
            }
            if (currentIndex == index && currentNode != null) {
                previousNode!!.next = currentNode.next
            }
        }
    }

}